### PR TITLE
Implemented keycloak authentication with Omniauth

### PR DESCRIPTION
**Pull request by RachalCassity**
Keycloak will be used to assign developer users roles and groups when entering the platform console.

For local development, developers can access keycloak by running: bin/keycloak.

To keep the platform console accessible to non-developers, users will be able to sign on via email/password.

Keycloak authentication process:
1) User clicks " Sign In Keycloak"
2) Console Platform and keycloak authentication 
  a) Platform sends an authentication request to keycloak with the parameters set in config/initializers/omniuath.rb
  b) User signs into keycloak
  c) Keycloak is redirected to the platform with a callback redirect_url. The production redirect url is configured in    Keycloak.
  d) The redirect url is routed to 'omniauth#keycloak_openid'. From the request environment headers, the uid, email, name, roles, and groups can be assigned to a new user or an existing user.

### CODE REVIEW

File: docs/README.md [hunk 1]
```diff
@@ -52,3 +52,22 @@ rails test test/path/to/the_test.rb:123
 ## Deployment
 
 While we await provisioning of appropriate services on the VA network, we are hosting the platform-console on Heroku. Each commit to master is automatically deployed and each pull request will generate a Review App.
+## KEYCLOAK INTEGRATION
```
Comment (ericboehs) [inline docs/README.md:55]: Could we extract this to a file in docs/?

File: db/migrate/20220131162210_add_oauth_parameters_to_users.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+# Added Oauth uid and provider to users.
+class AddOauthParametersToUsers < ActiveRecord::Migration[7.0]
+  def change
+    safety_assured { add_column :users, :uid, :string }
+    safety_assured { add_column :users, :provider, :string }
```
Comment (ericboehs) [inline db/migrate/20220131162210_add_oauth_parameters_to_users.rb:7]: I don't think we need to do safety_assured for just adding a column?

File: db/migrate/20220131162210_add_oauth_parameters_to_users.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+    safety_assured { remove_column :users, :password_digest, :string }
+    safety_assured { remove_column :users, :perishable_token, :string }
```
Comment (ericboehs) [inline db/migrate/20220131162210_add_oauth_parameters_to_users.rb:9]: perishable_token is how tests logged in without having to do the full login flow. I wonder how we can replicate that. 🤔

File: Gemfile [hunk 4]
```diff
@@ -33,6 +35,7 @@ end
 group :development, :test do
   gem 'brakeman'          # Detects security vulnerabilities in via static analysis
   gem 'bundler-audit'     # Provides patch-level verification for Bundled apps
+  gem 'dotenv-rails'
```
Comment (ericboehs) [inline Gemfile:38]: What do we need dotenv for here?

File: docs/README.md (hunk 1 above)
Comment (RachalCassity) [reply docs/README.md:55]: Yes, we definitely can.

File: db/migrate/20220131162210_add_oauth_parameters_to_users.rb (hunk 3 above)
Comment (RachalCassity) [reply db/migrate/20220131162210_add_oauth_parameters_to_users.rb:9]: That's the issue I am having. I'll keep perishable token.

**Comment by ericboehs** [comment]
Would we be able to use this? https://github.blog/changelog/2022-01-25-github-actions-reusable-workflows-can-be-referenced-locally/

File: app/controllers/sessions_controller.rb [hunk 5]
```diff
@@ -5,16 +5,17 @@ class SessionsController < ApplicationController
...
-    return create if params[:uid]
+    return create if params.dig('request.omniauth', :uid)
     return redirect_to root_path if current_user
 
     @user = User.new
   end
 
   def create
-    @user = User.find_by email: params[:email]
+    @result = true
```
Comment (ericboehs) [inline app/controllers/sessions_controller.rb:15]: This doesn't feel clean.

File: docs/README.md (hunk 1 above)
Comment (ericboehs) [reply docs/README.md:55]: Please do. :D

File: app/controllers/sessions_controller.rb (hunk 5 above)
Comment (ericboehs) [reply app/controllers/sessions_controller.rb:15]: Just noticed this is still marked WIP.

File: app/models/user.rb [hunk 6]
```diff
@@ -8,7 +8,18 @@ class User < ApplicationRecord
...
+
+  def self.from_omniauth(auth_hash)
+    User.find_or_initialize_by(uid: auth_hash['uid']) do |u|
+      u.name = auth_hash['info']['name']
+      u.email = auth_hash['info']['email']
+      u.password = SecureRandom.uuid
+      u.save!
+    end
+    # Authorize user and ensure keycloak is the provider
+    # teams = auth_hash['extra']['raw_info']['groups']
```
[the remark this answers is not in the document]
Comment (RachalCassity) [reply app/models/user.rb:20]: Yes, we will need this later on.

[the remark this answers is not in the document]
Comment (RachalCassity) [reply app/models/user.rb:20]: I commented them out because we do not have a team/users join name yet.

**Review by LindseySaari — APPROVED** [review]

**Pull request merged**